pydantic/main.py

```diff
@@ -866,12 +904,41 @@ def __eq__(self, other: Any) -> bool:
             self_type = self.__pydantic_generic_metadata__['origin'] or self.__class__
             other_type = other.__pydantic_generic_metadata__['origin'] or other.__class__
 
-            return (
+            # Perform common checks first
+            if not (
                 self_type == other_type
-                and self.__dict__ == other.__dict__
                 and self.__pydantic_private__ == other.__pydantic_private__
                 and self.__pydantic_extra__ == other.__pydantic_extra__
-            )
+            ):
+                return False
+
+            # Fix GH-7444 by comparing only pydantic fields
+            # We provide a fast-path for performance: __dict__ comparison is *much* faster
+            # See tests/benchmarks/test_basemodel_eq_performances.py and GH-7825 for benchmarks
+            if self.__dict__ == other.__dict__:
+                # If the check above passes, then pydantic fields are equal, we can return early
+                return True
+            else:
+                # Else, we need to perform a more detailed, costlier comparison
+                # We use operator.itemgetter because it is much faster than dict comprehensions
+                # NOTE: Contratry to standard python class and instances, when the Model class has
+                #       attribute default values and the model instance doesn't has a corresponding
+                #       attribute, accessing the missing attribute raises an error in
+                #       __getattr__ instance of returning the class attribute
+                #      Thus, using operator.itemgetter() instead of operator.attrgetter() is valid
+                model_fields = type(self).model_fields.keys()
+                getter = operator.itemgetter(*model_fields) if model_fields else lambda _: _SENTINEL
```

Comment:
We could do that, but I'm not sure it helps much. I'll wait on the feedbacl from @lig

Reply:
An easy path could be to use a function to retrieve the getter and a `functools.lru_cache` to keep the getter alive? Given the getter just stores the passed tuple, I'm not sure it would be faster

Reply:
Now that `model_fields` is also being used for the latest fast path, I think it's safe to leave it.